File: algorithms/mergesort.java
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class mergesort {
    public static void main(String[] args) throws IOException {

        final int LIST_LEN = 3000;

        File time_file = new File(args[2]);
        FileWriter timeWriter = new FileWriter(time_file);

        timeWriter.write("index,time\n");

        for (int iter = 0; iter < 1000; iter++) {
            List<Integer> original_list = new ArrayList<>();
            File infile = new File(args[0]);
            Scanner fileReader = new Scanner(infile);
            while (fileReader.hasNext()) {
                original_list.add(fileReader.nextInt());
            }
            fileReader.close();

            List<Integer> list_to_sort = new ArrayList<>(original_list);

            long starttime = System.nanoTime();

            List<Integer> buf_list = new ArrayList<>(original_list);

            int sectionlen = 1;
            while (sectionlen < LIST_LEN) {
                for (int sectionstart = 0; sectionstart < LIST_LEN; sectionstart += 2 * sectionlen) {
                    int index1 = sectionstart;
                    int index2 = sectionstart + sectionlen;
                    for (int i = sectionstart; i < Math.min(sectionstart + 2 * sectionlen, LIST_LEN); i++) {
                        if (index2 >= Math.min(sectionstart + 2 * sectionlen, LIST_LEN)) {
                            buf_list.set(i, list_to_sort.get(index1));
                            index1++;
                        } else if (index1 >= sectionstart + sectionlen) {
                            buf_list.set(i, list_to_sort.get(index2));
                            index2++;
                        } else if (list_to_sort.get(index1) < list_to_sort.get(index2)) {
                            buf_list.set(i, list_to_sort.get(index1));
                            index1++;
                        } else {
                            buf_list.set(i, list_to_sort.get(index2));
                            index2++;
                        }

                    }
                }
                sectionlen *= 2;
                List<Integer> temp = buf_list;
                buf_list = list_to_sort;
                list_to_sort = temp;
            }

            long endtime = System.nanoTime();

            timeWriter.write(Integer.toString(iter) + "," + Long.toString(endtime - starttime) + "\n");

            File answerFile = new File(args[1]);
            FileWriter ansWriter = new FileWriter(answerFile);
            for (int element : list_to_sort) {
                ansWriter.write(Integer.toString(element) + " ");
            }
            ansWriter.write("\n");
            ansWriter.close();
        }

        timeWriter.close();
    }
}

File: algorithms/mergesort.py
import sys
import time

LIST_LEN = 3000

time_file = open(sys.argv[3], "w")
time_file.write("index,time\n")

for iter in range(1000):
    original_list = []
    with open(sys.argv[1]) as input_file:
        for i in range(LIST_LEN):
            original_list.append(int(input_file.readline()))
    list_to_sort = original_list.copy()

    starttime = time.time_ns()

    buf_list = [None] * len(list_to_sort)

    sectionlen = 1
    while sectionlen < len(list_to_sort):
        for sectionstart in range(0, len(list_to_sort), 2 * sectionlen):
            index1 = sectionstart
            index2 = sectionstart + sectionlen
            for i in range(sectionstart, min(sectionstart + 2 * sectionlen, len(list_to_sort))):
                if index2 >= min(sectionstart + 2 * sectionlen, len(list_to_sort)):
                    buf_list[i] = list_to_sort[index1]
                    index1 += 1
                elif index1 >= sectionstart + sectionlen:
                    buf_list[i] = list_to_sort[index2]
                    index2 += 1
                elif list_to_sort[index1] < list_to_sort[index2]:
                    buf_list[i] = list_to_sort[index1]
                    index1 += 1
                else:
                    buf_list[i] = list_to_sort[index2]
                    index2 += 1

        sectionlen *= 2
        buf_list, list_to_sort = list_to_sort, buf_list

    endtime = time.time_ns()
    time_file.write(str(iter) + "," + str(endtime - starttime) + "\n")

    with open(sys.argv[2], "w") as answer_file:
        for element in list_to_sort:
            answer_file.write(str(element) + " ")
        answer_file.write("\n")
time_file.close()
